### PR TITLE
Tests for Eject Button

### PR DESCRIPTION
**Pull request by AlexOn1ine**
## Description
Added tests for Eject Button. The source for the tests was bulbagarden.
One additional test for intimidate was added to test a bug.

### CODE REVIEW

**Comment by SBird1337** [comment]
I guess this
```
Intimidate and Eject Button force the opponent to Attack: FAIL
```
is supposed to fail / known to fail, in which case you need to supply `KNOWN_FAILING` (in the best case with a corresponding issue / pr number) - otherwise the CI won't pass.

**Comment by AlexOn1ine** [comment]
> is supposed to fail / known to fail, in which case you need to supply `KNOWN_FAILING` (in the best case with a corresponding issue / pr number) - otherwise the CI won't pass.

Like this?
image

**Comment by AlexOn1ine** [comment]
Added `KNOWN_FAILING` to the Intimidate test.

File: test/ability_intimidate.c [hunk 1]
```diff
@@ -95,3 +95,30 @@ DOUBLE_BATTLE_TEST("Intimidate doesn't activate on an empty field in a double ba
...
 }
+
+SINGLE_BATTLE_TEST("Intimidate and Eject Button force the opponent to Attack")
+{
+    KNOWN_FAILING; // Issue #2837
+    GIVEN {
+        ASSUME(gItems[ITEM_EJECT_BUTTON].holdEffect == HOLD_EFFECT_EJECT_BUTTON);
+        PLAYER(SPECIES_WOBBUFFET)
+        OPPONENT(SPECIES_WOBBUFFET) { Item(ITEM_EJECT_BUTTON); };
+        OPPONENT(SPECIES_HITMONTOP) { Moves(MOVE_TACKLE); };
```
Comment (SBird1337) [inline test/ability_intimidate.c:106]:
```suggestion
        PLAYER(SPECIES_WOBBUFFET);
        OPPONENT(SPECIES_WOBBUFFET) { Item(ITEM_EJECT_BUTTON); };
        OPPONENT(SPECIES_HITMONTOP);
```

File: test/ability_intimidate.c [hunk 2]
```diff
@@ -95,3 +95,30 @@ DOUBLE_BATTLE_TEST("Intimidate doesn't activate on an empty field in a double ba
...
+    } WHEN {
+        TURN { MOVE(player, MOVE_QUICK_ATTACK);
+               MOVE(opponent, MOVE_TACKLE);
+               SEND_OUT(opponent, 1);
+        }
```
Comment (SBird1337) [inline test/ability_intimidate.c:111]:
```suggestion
        TURN { 
               MOVE(player, MOVE_QUICK_ATTACK);
               MOVE(opponent, MOVE_TACKLE);
               SEND_OUT(opponent, 1);
        }
```

File: test/hold_effect_eject_button.c [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+{
+    ASSUME(gItems[ITEM_EJECT_BUTTON].holdEffect == HOLD_EFFECT_EJECT_BUTTON);
+}
+
+SINGLE_BATTLE_TEST("Eject Button is not triggered when there is nothing to switch in")
+{
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_WOBBUFFET) { Item(ITEM_EJECT_BUTTON); }
+        PLAYER(SPECIES_WOBBUFFET) { HP(0); }
```
Comment (SBird1337) [inline test/hold_effect_eject_button.c:14]: Is there a specific reason the player has 0 HP here?

Comment (SBird1337) [inline test/hold_effect_eject_button.c:14]:
```suggestion
        OPPONENT(SPECIES_WOBBUFFET) { Item(ITEM_EJECT_BUTTON); };
        PLAYER(SPECIES_WOBBUFFET) { HP(0); };
```

File: test/hold_effect_eject_button.c [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        }
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_TACKLE, opponent);
+    }
+}
+
+SINGLE_BATTLE_TEST("Eject Button is not activated by a Sheer Force boosted move")
+{
+    GIVEN {
+        PLAYER(SPECIES_NIDOKING) { Ability(ABILITY_SHEER_FORCE); }
+        OPPONENT(SPECIES_WOBBUFFET) { Item(ITEM_EJECT_BUTTON); }
```
Comment (SBird1337) [inline test/hold_effect_eject_button.c:34]:
```suggestion
        PLAYER(SPECIES_NIDOKING) { Ability(ABILITY_SHEER_FORCE); };
        OPPONENT(SPECIES_WOBBUFFET) { Item(ITEM_EJECT_BUTTON); };
```

File: test/hold_effect_eject_button.c [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        }
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_TACKLE, opponent);
+    }
+}
+
+SINGLE_BATTLE_TEST("Eject Button will not activate under Substitute")
+{
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_RAICHU) { Item(ITEM_EJECT_BUTTON); }
```
Comment (SBird1337) [inline test/hold_effect_eject_button.c:55]:
```suggestion
        OPPONENT(SPECIES_RAICHU) { Item(ITEM_EJECT_BUTTON); };
```

File: test/hold_effect_eject_button.c [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(opponent, MOVE_SUBSTITUTE);
+               MOVE(player, MOVE_TACKLE);
```
Comment (SBird1337) [inline test/hold_effect_eject_button.c:59]:
```suggestion
        TURN { 
               MOVE(opponent, MOVE_SUBSTITUTE);
               MOVE(player, MOVE_TACKLE);
```

File: test/hold_effect_eject_button.c [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+            MESSAGE("Foe Raichu is switched out with the Eject Button!");
+        }
+    }
+}
+
+SINGLE_BATTLE_TEST("Eject Button is not blocked by trapping abilities or moves")
+{
+    GIVEN {
+        PLAYER(SPECIES_DUGTRIO) { Ability(ABILITY_ARENA_TRAP); }
+        OPPONENT(SPECIES_WOBBUFFET) { Item(ITEM_EJECT_BUTTON); }
```
Comment (SBird1337) [inline test/hold_effect_eject_button.c:77]:
```suggestion
        PLAYER(SPECIES_DUGTRIO) { Ability(ABILITY_ARENA_TRAP); };
        OPPONENT(SPECIES_WOBBUFFET) { Item(ITEM_EJECT_BUTTON); };
```

File: test/hold_effect_eject_button.c [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+        MESSAGE("2 sent out Wobbuffet!");
+    }
+}
+
+SINGLE_BATTLE_TEST("Eject Button is not triggered after the mon loses Eject Button")
+{
+    GIVEN {
+        PLAYER(SPECIES_RAICHU);
+        OPPONENT(SPECIES_WOBBUFFET) { Item(ITEM_EJECT_BUTTON); }
+        OPPONENT(SPECIES_WOBBUFFET);
```
Comment (SBird1337) [inline test/hold_effect_eject_button.c:97]:
```suggestion
        OPPONENT(SPECIES_WOBBUFFET) { Item(ITEM_EJECT_BUTTON); };
        OPPONENT(SPECIES_WOBBUFFET);
```

File: test/hold_effect_eject_button.c [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    }
+}
+
+SINGLE_BATTLE_TEST("Eject Button is not triggered after given to player by Picketpocket")
+{
+    GIVEN {
+        PLAYER(SPECIES_REGIELEKI) { Item(ITEM_EJECT_BUTTON); }
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_SNEASEL) { Ability(ABILITY_PICKPOCKET); }
+        OPPONENT(SPECIES_WOBBUFFET);
```
Comment (SBird1337) [inline test/hold_effect_eject_button.c:119]:
```suggestion
        PLAYER(SPECIES_REGIELEKI) { Item(ITEM_EJECT_BUTTON); };
        PLAYER(SPECIES_WOBBUFFET);
        OPPONENT(SPECIES_SNEASEL) { Ability(ABILITY_PICKPOCKET); };
        OPPONENT(SPECIES_WOBBUFFET);
```

File: test/hold_effect_eject_button.c [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+        MESSAGE("Foe Sneasel stole Regieleki's Eject Button!");
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_TACKLE, opponent);
+    }
+}
+
+SINGLE_BATTLE_TEST("Eject Button has no chance to activate after Dragon Tai or Circle Throw")
+{
+    GIVEN {
+        PLAYER(SPECIES_KOMMO_O);
+        OPPONENT(SPECIES_WOBBUFFET) { Item(ITEM_EJECT_BUTTON); }
```
Comment (SBird1337) [inline test/hold_effect_eject_button.c:137]:
```suggestion
        OPPONENT(SPECIES_WOBBUFFET) { Item(ITEM_EJECT_BUTTON); };
```

File: test/hold_effect_eject_button.c [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+        }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_TACKLE, player);
+        ABILITY_POPUP(opponent, ABILITY_PICKPOCKET);
+        MESSAGE("Foe Sneasel stole Regieleki's Eject Button!");
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_TACKLE, opponent);
+    }
+}
+
+SINGLE_BATTLE_TEST("Eject Button has no chance to activate after Dragon Tai or Circle Throw")
```
Comment (SBird1337) [inline test/hold_effect_eject_button.c:133]:
```suggestion
SINGLE_BATTLE_TEST("Eject Button has no chance to activate after Dragon Tail")
```
the test does not actually involve Circle Throw.

File: test/hold_effect_eject_button.c [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+        OPPONENT(SPECIES_WOBBUFFET) { Item(ITEM_EJECT_BUTTON); }
+        OPPONENT(SPECIES_CHANSEY);
+    } WHEN {
+        TURN {
+            MOVE(player, MOVE_DRAGON_TAIL);
+            MOVE(opponent, MOVE_TACKLE);
+        }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_DRAGON_TAIL, player);
+        MESSAGE("Foe Chansey was dragged out!");
```
Comment (SBird1337) [inline test/hold_effect_eject_button.c:146]: shouldnt this check for NONE_OF(...) for eject button again?

**Review by SBird1337 — CHANGES_REQUESTED** [review]

**Comment by AlexOn1ine** [comment]
> I went ahead and comitted the suggestions about missing semicolons and the typo, hope you don't mind :)

All good. Thanks for mentioning it.

File: test/hold_effect_eject_button.c (hunk 3 above)
Comment (AlexOn1ine) [reply test/hold_effect_eject_button.c]: So the AI isn't switching a dead mon to battle.

File: test/ability_intimidate.c [hunk 13]
```diff
@@ -95,3 +95,31 @@ DOUBLE_BATTLE_TEST("Intimidate doesn't activate on an empty field in a double ba
...
 }
+
+SINGLE_BATTLE_TEST("Intimidate and Eject Button force the opponent to Attack")
+{
+    KNOWN_FAILING; // Issue #2837
+    GIVEN {
+        ASSUME(gItems[ITEM_EJECT_BUTTON].holdEffect == HOLD_EFFECT_EJECT_BUTTON);
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_WOBBUFFET) { Item(ITEM_EJECT_BUTTON); };
+        OPPONENT(SPECIES_HITMONTOP);
```
Comment (SBird1337) [inline test/ability_intimidate.c:106]:
```suggestion
        OPPONENT(SPECIES_HITMONTOP) { Moves(MOVE_TACKLE); };
```

**Review by SBird1337 — APPROVED** [review]

**Pull request merged**